### ISSUE TITLE
Dialog Component Role Attribute Type Error

### ISSUE THREAD

**Issue by szkieb** [tailwindlabs/headlessui#2508]
**What package within Headless UI are you using?**
    "@headlessui/react": "1.7.14", Dialog Component

**What version of that package are you using?**
    "@headlessui/react": "1.7.14",

**What browser are you using?**
Firefox

**Reproduction URL**
https://codesandbox.io/s/currying-silence-3pu19w?file=/src/App.tsx

**Describe your issue**
Using the Dialog Compoent from Headless UI, I get a typescript error when using the role attribute

>Type '{ children: Element; role: string; open: boolean; onClose: Dispatch<SetStateAction<boolean>>; }' is not assignable to type 'IntrinsicAttributes & (DialogProps<"div"> & { ref?: Ref<HTMLDivElement> | undefined; })'.
  Property 'role' does not exist on type 'IntrinsicAttributes & (DialogProps<"div"> & { ref?: Ref<HTMLDivElement> | undefined; })'.ts(2322)

The issue did not occur in earlier versions, see: https://codesandbox.io/s/laughing-edison-182d99?file=/package.json

**Comment by thecrypticace** [comment]
Hey! So the types are correct here. This was a change we introduced in 1.7.12 which made the types more accurate. We control the `role` prop for our components. In the case of `<Dialog>` it is always `dialog`. Any change to it previously did not take effect — but the types did not indicate this (see screenshot below). When we updated the types to be more accurate, it started showing the error. This indicates that `role` is not a valid prop to pass and cannot be changed — as we set it for you.

<img width="839" alt="Screenshot 2023-05-23 at 09 39 10" src="https://github.com/tailwindlabs/headlessui/assets/614993/b53b16f4-e620-4e17-b8dd-af4df85ac466">

Hope that helps! ✨
